raster-foundry/raster-foundry
# Upgrade circe and maml and upgrade async jobs to use IOApp

## Overview

This PR upgrades circe and maml in hopes that that fixes assembly throwing some of our json decoders in the garbage in the `batch` subproject jar. It does!

It also works on job termination by:

- making sure we only try to exit the job once
- making sure we're explicit about set up and teardown of any resources by enclosing them in cats effect's `Resource`s (spark contexts, threadpools, and transactors)

### Checklist


### Notes

There were a few cases where specific exceptions became general exceptions, which isn't great, but exceptions expressing those errors have been dropped from maml, so :man_shrugging:.

Also, I can't get lab tiles or run exports :tada: and I'm not sure why. My exports are blowing up with memory errors and lab tiles keep failing with socket timeouts. If this happens for you as well, I'll test from some other network to confirm whether it's related to changes in this PR or some State of the Universe :tm: problem.

## Testing Instructions

 * Make sure you can still get lab tiles from the non-backsplash tile server
 * Make sure you can run exports
 * Run `rf update-aoi-project <some aoi project id>` and make sure you actually start doing some work instead of exploding with some kind of circe decoder not found error
 * Run `java -cp /opt/raster-foundry/jars/batch-assembly.jar com.rasterfoundry.batch.Main import_landsat8_c1` and make sure it terminates
 * delete all (or at least some of) the cog histograms from your database and run `java -cp /opt/raster-foundry/jars/batch-assembly.jar com.rasterfoundry.batch.Main cog-histogram-backfill` and make sure it terminates

Closes #4291
